internal/app/init/pkg/system/services/kubeadm/kubeadm.go

```diff
@@ -0,0 +1,304 @@
+			KeyPath:  constants.KubeadmFrontProxyCAKey,
+		},
+		{
+			Cert:     data.Security.Kubernetes.Etcd,
+			CertPath: constants.KubeadmEtcdCACert,
+			KeyPath:  constants.KubeadmEtcdCAKey,
+		},
+	}
+
+	for _, cert := range certs {
+		if cert.Cert == nil {
+			continue
+		}
+
+		if cert.Cert.Crt != nil {
+			if err = os.MkdirAll(path.Dir(cert.CertPath), dirPerm); err != nil {
+				return err
+			}
+			if err = ioutil.WriteFile(cert.CertPath, cert.Cert.Crt, certPerm); err != nil {
+				return fmt.Errorf("write %s: %v", cert.CertPath, err)
+			}
+		}
+		if cert.Cert.Key != nil {
+			if err = os.MkdirAll(path.Dir(cert.KeyPath), dirPerm); err != nil {
+				return err
+			}
+			if err = ioutil.WriteFile(cert.KeyPath, cert.Cert.Key, certPerm); err != nil {
+				return fmt.Errorf("write %s: %v", cert.KeyPath, err)
+			}
+		}
+	}
+	return err
+}
+
+// RequiredFiles returns a slice of the required CA and
+// security policies necessary for kubeadm init to function.
+// This serves as a base for the list of files that need to
+// be synced via trustd from other nodes
+func RequiredFiles() []string {
+	return []string{
+		constants.AuditPolicyPathInitramfs,
+		constants.EncryptionConfigInitramfsPath,
+		constants.KubeadmCACert,
+		constants.KubeadmCAKey,
+		constants.KubeadmSACert,
+		constants.KubeadmSAKey,
+		constants.KubeadmFrontProxyCACert,
+		constants.KubeadmFrontProxyCAKey,
+		constants.KubeadmEtcdCACert,
+		constants.KubeadmEtcdCAKey,
+	}
+}
+
+// FileSet compares the list of required files to the ones
+// already present on the node and returns the delta
+func FileSet() []*proto.ReadFileRequest {
+	files := RequiredFiles()
+
+	fileRequests := []*proto.ReadFileRequest{}
+	// Check to see if we already have the file locally
+	for _, file := range files {
+		if _, err := os.Stat(file); os.IsNotExist(err) {
+			fileRequests = append(fileRequests, &proto.ReadFileRequest{Path: file})
+		}
+	}
+
+	return fileRequests
+}
+
+// CreateTrustdClients handles instantiating a trustd client connection
+// to each trustd endpoint defined in userdata
+func CreateTrustdClients(data *userdata.UserData) ([]proto.TrustdClient, error) {
+	var creds basic.Credentials
+	var err error
+
+	trustds := []proto.TrustdClient{}
+
+	creds, err = basic.NewCredentials(data.Services.Trustd)
+	if err != nil {
+		return trustds, err
+	}
+
+	// Create a trustd client for each endpoint to set up
+	// a fan out approach to gathering the files
+	var conn *grpc.ClientConn
+	for _, endpoint := range data.Services.Trustd.Endpoints {
+		conn, err = basic.NewConnection(endpoint, constants.TrustdPort, creds)
+		if err != nil {
+			return trustds, err
+		}
+		trustds = append(trustds, proto.NewTrustdClient(conn))
+	}
+	return trustds, nil
+}
+
+// Download handles the retrieval of files from a trustd endpoint
+func Download(ctx context.Context, client proto.TrustdClient, file *proto.ReadFileRequest, content chan<- []byte) {
+	select {
+	case <-ctx.Done():
+		return
```

Comment:
you can probably remove these returns, as it's going to return anyways on any code path ;)